### DISCUSSION TITLE
Lowest common denominator of set and fronzenset?

### DISCUSSION THREAD

**Discussion by dimaqq** [microsoft/pyright#9591] [Q&A] [answered]
How to type a function argument so that both set and frozenset can be passed?
I mean to make a promise to the function user that:
- function doesn't modify the argument
- function doesn't need it to be hashable

(To be fair this is really a form of "repeatable [unordered] iterable" for which there's no good type hint).

**Comment by erictraut** [comment]
I'm not sure I understand your question. Would a simple union type meet your needs?

```python
def func(x: set[Any] | frozenset[Any]):
    pass
```

**Reply by dimaqq** [reply]
If that’s the best possible, sure, though it gets rather repetitive when parametrized.

**Reply by erictraut** [reply]
You can create a type alias to reduce repetition:

```python
type MySet = set[Any] | frozenset[Any]

def func(x: MySet):
    pass
```

**Comment by Azureblade3808** [comment, marked as answer]
`collections.abc.Set` (also known as `typing.AbstractSet`) may suit your need.

```python
from collections.abc import Set

def func(x: Set[int]): ...

func({0})
func(frozenset({0}))
```
